File: DoAnChuyenNganh-main/DoAnCoSo/WebDuLichDaLat/Services/CarpoolMatchingService.cs
using WebDuLichDaLat.Models;
using System.Collections.Generic;
using System.Linq;

namespace WebDuLichDaLat.Services
{
    /// <summary>
    /// Dich vu tinh toan va ghep noi hanh khach voi phuong tien (Carpool Matching).
    /// Su dung ket hop cac thuat toan K-Means Cluster, Min-Cost Max-Flow va PDPTW de toi uu hoa lo trinh.
    /// </summary>
    public class CarpoolMatchingService
    {
        private readonly KMeansClusteringService _kmeansService;
        private readonly MinCostMaxFlowService _flowService;
        private readonly PDPTWService _pdptwService;

        public CarpoolMatchingService()
        {
            _kmeansService = new KMeansClusteringService();
            _flowService = new MinCostMaxFlowService();
            _pdptwService = new PDPTWService();
        }

        /// <summary>
        /// Thuc hien ghep noi danh sach hanh khach vao cac phuong tien hien co.
        /// </summary>
        /// <param name="vehicles">Danh sach xe kha dung</param>
        /// <param name="passengers">Danh sach hanh khach cho ghep</param>
        /// <param name="groups">Danh sach nhom hanh khach</param>
        /// <returns>Danh sach ket qua ghep noi kem lo trinh toi uu</returns>
        public List<CarpoolMatchResult> MatchPassengersToVehicles(
            List<Vehicle> vehicles,
            List<Passenger> passengers,
            List<PassengerGroup> groups)
        {
            var results = new List<CarpoolMatchResult>();

            if (!vehicles.Any() || !passengers.Any())
                return results;

            // Su dung thuat toan K-Means de phan cum hanh khach theo vi tri dia ly
            int k = Math.Min(5, Math.Max(1, passengers.Count / 3));
            var clusters = _kmeansService.ClusterPassengers(passengers, k);

            // Giai bai toan ghep noi bang thuat toan Min-Cost Max-Flow
            var flowResult = _flowService.Solve(vehicles, passengers, groups);

            // Toi uu hoa lo trinh di chuyen cho tung phuong tien sau khi da ghep noi
            foreach (var assignment in flowResult.Assignments)
            {
                int vehicleId = assignment.Key;
                var vehicle = vehicles.FirstOrDefault(v => v.Id == vehicleId);
                if (vehicle == null) continue;

                // Khoi tao danh sach hanh khach gan cho xe theo ket qua tu luong (flow)
                var rawVehiclePassengers = passengers
                    .Where(p => assignment.Value.Contains(p.Id))
                    .ToList();

                // Kiem tra va gioi han so luong hanh khach dua tren suc chua thuc te cua phuong tien
                var vehiclePassengers = new List<Passenger>();
                int usedSeats = 0;

                foreach (var p in rawVehiclePassengers)
                {
                    int seatsForPassenger = 1;

                    if (p.GroupId.HasValue)
                    {
                        var g = groups.FirstOrDefault(x => x.Id == p.GroupId.Value);
                        seatsForPassenger = g?.RequiredSeats ?? 1;
                    }

                    // Kiem tra dieu kien suc chua: Neu vuot qua gioi han, hanh khach nay se khong duoc xep vao xe hien tai
                    if (usedSeats + seatsForPassenger > vehicle.TotalSeats)
                        continue;

                    vehiclePassengers.Add(p);
                    usedSeats += seatsForPassenger;
                }

                if (!vehiclePassengers.Any())
                    continue;

                // Thuc hien toi uu hoa lo trinh don/tra khach (Pickup and Delivery Problem)
                var optimizedRoute = _pdptwService.OptimizeRoute(vehicle, vehiclePassengers);


                // Tính tổng chi phí
                decimal totalCost = vehicle.FixedPrice.HasValue && vehicle.FixedPrice.Value > 0
                    ? vehicle.FixedPrice.Value
                    : optimizedRoute.TotalCost;

                // Tổng ghế khách thực sự sử dụng (đã giới hạn)
                int occupiedSeats = usedSeats;

                // Số ghế trống
                int availableSeats = Math.Max(0, vehicle.TotalSeats - occupiedSeats);

                // Chi phí mỗi ghế
                decimal costPerSeat = occupiedSeats > 0
                    ? totalCost / occupiedSeats
                    : 0;

                // Tạo kết quả
                var matchResult = new CarpoolMatchResult
                {
                    VehicleId = vehicle.Id,
                    DriverName = vehicle.DriverName,
                    LicensePlate = vehicle.LicensePlate,
                    VehicleType = vehicle.VehicleType,
                    TotalSeats = vehicle.TotalSeats,
                    AvailableSeats = availableSeats,     // ĐÃ FIX
                    OccupiedSeats = occupiedSeats,       // ĐÃ FIX
                    TotalCost = totalCost,
                    CostPerPassenger = costPerSeat,
                    TotalDistance = optimizedRoute.TotalDistance,
                    EstimatedDepartureTime = optimizedRoute.StartTime,
                    EstimatedArrivalTime = optimizedRoute.EndTime,
                    PickupAddress = vehicle.PickupAddress ?? "",
                    DropoffAddress = vehicle.DropoffAddress ?? ""
                };

                // Danh sách matched passengers
                int sequence = 1;

                foreach (var node in optimizedRoute.Nodes)
                {
                    if (node.PassengerId.HasValue)
                    {
                        var passenger = vehiclePassengers
                            .FirstOrDefault(p => p.Id == node.PassengerId.Value);

                        if (passenger != null)
                        {
                            var existingMatch = matchResult.MatchedPassengers
                                .FirstOrDefault(m => m.PassengerId == passenger.Id);

                            if (existingMatch == null)
                            {
                                int seatsForPassenger = 1;

                                if (passenger.GroupId.HasValue)
                                {
                                    var g = groups.FirstOrDefault(x => x.Id == passenger.GroupId.Value);
                                    seatsForPassenger = g?.RequiredSeats ?? 1;
                                }

                                decimal passengerCost = seatsForPassenger * costPerSeat;

                                existingMatch = new PassengerMatch
                                {
                                    PassengerId = passenger.Id,
                                    PassengerName = passenger.Name,
                                    PickupAddress = passenger.PickupAddress ?? "",
                                    DropoffAddress = passenger.DropoffAddress ?? "",
                                    Cost = passengerCost,
                                    SequenceOrder = sequence++
                                };

                                matchResult.MatchedPassengers.Add(existingMatch);
                            }

                            if (node.Type == "pickup")
                                existingMatch.PickupTime = node.EarliestTime;

                            if (node.Type == "dropoff")
                                existingMatch.DropoffTime = node.LatestTime;
                        }
                    }
                }

                // Lưu route points
                sequence = 1;
                foreach (var node in optimizedRoute.Nodes)
                {
                    matchResult.OptimizedRoute.Add(new RoutePoint
                    {
                        Latitude = node.Latitude,
                        Longitude = node.Longitude,
                        Address = node.Address,
                        Type = node.Type,
                        PassengerId = node.PassengerId,
                        Time = node.EarliestTime,
                        Sequence = sequence++
                    });
                }

                results.Add(matchResult);
            }

            return results;
        }
    }
}

File: src/main/java/com/example/dalattravel/service/CarpoolMatchingService.java
package com.example.DaLattravel.service;

import com.example.DaLattravel.dto.*;
import com.example.DaLattravel.model.Passenger;
import com.example.DaLattravel.model.PassengerGroup;
import com.example.DaLattravel.model.Vehicle;
import org.springframework.stereotype.Service;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.util.*;
import java.util.stream.Collectors;

@Service
public class CarpoolMatchingService {

    private final KMeansClusteringService kmeansService;
    private final MinCostMaxFlowService flowService;
    private final PDPTWService pdptwService;

    public CarpoolMatchingService(KMeansClusteringService kmeansService, MinCostMaxFlowService flowService, PDPTWService pdptwService) {
        this.kmeansService = kmeansService;
        this.flowService = flowService;
        this.pdptwService = pdptwService;
    }

    public List<CarpoolMatchResult> matchPassengersToVehicles(
            List<Vehicle> vehicles,
            List<Passenger> passengers,
            List<PassengerGroup> groups) {

        List<CarpoolMatchResult> results = new ArrayList<>();

        if (vehicles.isEmpty() || passengers.isEmpty()) {
            return results;
        }

        int k = Math.min(5, Math.max(1, passengers.size() / 3));
        kmeansService.clusterPassengers(passengers, k);

        MinCostMaxFlowService.FlowResult flowResult = flowService.solve(vehicles, passengers, groups);

        for (Map.Entry<Integer, List<Integer>> assignment : flowResult.getAssignments().entrySet()) {
            int vehicleId = assignment.getKey();
            Vehicle vehicle = vehicles.stream().filter(v -> v.getId() == vehicleId).findFirst().orElse(null);
            if (vehicle == null) continue;

            List<Passenger> rawVehiclePassengers = passengers.stream()
                    .filter(p -> assignment.getValue().contains(p.getId()))
                    .collect(Collectors.toList());

            List<Passenger> vehiclePassengers = new ArrayList<>();
            int usedSeats = 0;

            for (Passenger p : rawVehiclePassengers) {
                int seatsForPassenger = 1;

                if (p.getGroupId() != null) {
                    PassengerGroup g = groups.stream().filter(x -> Objects.equals(x.getId(), p.getGroupId())).findFirst().orElse(null);
                    seatsForPassenger = g != null ? g.getRequiredSeats() : 1;
                }

                if (usedSeats + seatsForPassenger > vehicle.getTotalSeats()) continue;

                vehiclePassengers.add(p);
                usedSeats += seatsForPassenger;
            }

            if (vehiclePassengers.isEmpty()) continue;

            PDPTWRoute optimizedRoute = pdptwService.optimizeRoute(vehicle, vehiclePassengers);

            BigDecimal totalCost = (vehicle.getFixedPrice() != null && vehicle.getFixedPrice().compareTo(BigDecimal.ZERO) > 0)
                    ? vehicle.getFixedPrice()
                    : optimizedRoute.getTotalCost();

            int occupiedSeats = usedSeats;
            int availableSeats = Math.max(0, vehicle.getTotalSeats() - occupiedSeats);

            BigDecimal costPerSeat = occupiedSeats > 0
                    ? totalCost.divide(BigDecimal.valueOf(occupiedSeats), 2, RoundingMode.HALF_UP)
                    : BigDecimal.ZERO;

            CarpoolMatchResult matchResult = CarpoolMatchResult.builder()
                    .vehicleId(vehicle.getId())
                    .driverName(vehicle.getDriverName())
                    .licensePlate(vehicle.getLicensePlate())
                    .vehicleType(vehicle.getVehicleType())
                    .totalSeats(vehicle.getTotalSeats())
                    .availableSeats(availableSeats)
                    .occupiedSeats(occupiedSeats)
                    .totalCost(totalCost)
                    .costPerPassenger(costPerSeat)
                    .totalDistance(optimizedRoute.getTotalDistance())
                    .estimatedDepartureTime(optimizedRoute.getStartTime())
                    .estimatedArrivalTime(optimizedRoute.getEndTime())
                    .pickupAddress(vehicle.getPickupAddress() != null ? vehicle.getPickupAddress() : "")
                    .dropoffAddress(vehicle.getDropoffAddress() != null ? vehicle.getDropoffAddress() : "")
                    .matchedPassengers(new ArrayList<>())
                    .optimizedRoute(new ArrayList<>())
                    .build();

            int sequence = 1;

            for (PDPTWNode node : optimizedRoute.getNodes()) {
                if (node.getPassengerId() != null) {
                    Passenger passenger = vehiclePassengers.stream()
                            .filter(p -> p.getId().equals(node.getPassengerId()))
                            .findFirst().orElse(null);

                    if (passenger != null) {
                        PassengerMatch existingMatch = matchResult.getMatchedPassengers().stream()
                                .filter(m -> m.getPassengerId() == passenger.getId())
                                .findFirst().orElse(null);

                        if (existingMatch == null) {
                            int seatsForPassenger = 1;

                            if (passenger.getGroupId() != null) {
                                PassengerGroup g = groups.stream().filter(x -> Objects.equals(x.getId(), passenger.getGroupId())).findFirst().orElse(null);
                                seatsForPassenger = g != null ? g.getRequiredSeats() : 1;
                            }

                            BigDecimal passengerCost = costPerSeat.multiply(BigDecimal.valueOf(seatsForPassenger));

                            existingMatch = PassengerMatch.builder()
                                    .passengerId(passenger.getId())
                                    .passengerName(passenger.getName())
                                    .pickupAddress(passenger.getPickupAddress() != null ? passenger.getPickupAddress() : "")
                                    .dropoffAddress(passenger.getDropoffAddress() != null ? passenger.getDropoffAddress() : "")
                                    .cost(passengerCost)
                                    .sequenceOrder(sequence++)
                                    .build();

                            matchResult.getMatchedPassengers().add(existingMatch);
                        }

                        if ("pickup".equals(node.getType()))
                            existingMatch.setPickupTime(node.getEarliestTime());

                        if ("dropoff".equals(node.getType()))
                            existingMatch.setDropoffTime(node.getLatestTime());
                    }
                }
            }

            sequence = 1;
            for (PDPTWNode node : optimizedRoute.getNodes()) {
                matchResult.getOptimizedRoute().add(RoutePoint.builder()
                        .latitude(node.getLatitude())
                        .longitude(node.getLongitude())
                        .address(node.getAddress())
                        .type(node.getType())
                        .passengerId(node.getPassengerId())
                        .time(node.getEarliestTime())
                        .sequence(sequence++)
                        .build());
            }

            results.add(matchResult);
        }

        return results;
    }
}
